File: v-lang/l-java/v-spring/spring-jpa/src/main/java/org/example/model/BaseModel.java
package org.example.model;

import jakarta.persistence.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.io.Serializable;
import java.time.Instant;

@MappedSuperclass
@EntityListeners(AuditingEntityListener.class)
public abstract class BaseModel implements Serializable {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    public Long id;

    @Column(name = "created_at", nullable = false, updatable = false)
    @CreatedDate
    public Instant createdAt;

    @Column(name = "updated_at")
    @LastModifiedDate
    public Instant updatedAt;

    @Column(name = "deleted_at")
    public Instant deletedAt;
}

File: v-lang/lang.kt/v-spring/spring-jpa/src/main/kotlin/org/example/model/BaseModel.kt
package org.example.model

import jakarta.persistence.*
import org.example.anno.NoArg
import org.hibernate.annotations.Where
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.time.Instant
import java.time.LocalDateTime

@NoArg
@MappedSuperclass
@EntityListeners(AuditingEntityListener::class)
abstract class BaseModel(

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    open val id: Long?,

    @Column(name = "created_at")
    @CreatedDate
    open val createdAt: Instant?,

    @Column(name = "updated_at")
    @LastModifiedDate
    open val updatedAt: Instant?,

    @Column(name = "deleted_at")
    open val deletedAt: Instant?,
)